### ISSUE TITLE
arXiv-2019/02-BERT for Joint Intent Classification and Slot Filling

### ISSUE THREAD

**Issue by BrambleXu** [BrambleXu/knowledge-graph-learning#126]
**一句话总结：**

结合了意图分类和填槽两个任务，基于BERT的联合模型。

**资源：**

- [pdf](https://arxiv.org/abs/1902.10909)
- [code](
- [paper-with-code](https://paperswithcode.com/paper/bert-for-joint-intent-classification-and-slot)

**关键字：**

- dataset: 
- Intent Classification
- Slot Filling
- [What is the difference between slot filling in NLU and named entity recognition in NLP?](https://www.reddit.com/r/LanguageTechnology/comments/45g5hr/what_is_the_difference_between_slot_filling_in/)
   - 简单的说，NER是SF的一个子任务。SF的范围更广。而且SF通常可以分为多step进行。

**笔记：**

智能语音助手，主要涉及到了NLU任务。Natural language understanding (NLU) is
critical to the performance of goal-oriented spoken dialogue systems。NLU主要包含了intent classification and slot filling tasks，目的是形成用户话语的语义解析.

- Intent classification focuses on predicting the intent of the query
- Slot filling extracts semantic concepts.

![image](https://user-images.githubusercontent.com/10768193/56557613-b5433c80-65d6-11e9-9331-e7be72c391ba.png)
（这个任务RE和NER的配置非常像）

- Intent classification: 分类任务。
- Slot filling: 序列标注任务。

Hence, attention-based joint learning methods were proposed and achieved the state-of-the-art performance for joint intent classification and slot filling (Liu and Lane, 2016; Goo et al., 2018). 看来基于attention的模型是SOTA。

另外为了解决标注数据缺失的问题，有ELMo和BERT这样的模型被提出。

这篇文章的贡献：
- 利用BERT提高NLU问题的表现力
- 提出了一个a joint intent classification and slot filling model based on BERT

![image](https://user-images.githubusercontent.com/10768193/56558212-e9b7f800-65d8-11e9-983b-fc0a8aacd816.png)

A special classification embedding ([CLS]) is inserted as the first token and a special
token ([SEP]) is added as the final token.

3.2 Joint Intent Classification and Slot Filling

![image](https://user-images.githubusercontent.com/10768193/56558389-84b0d200-65d9-11e9-924e-6c5d55d2f108.png)

分类问题，直接用[CLS]的embedding，接一个softmax就OK了。而对于slot filling, 需要用wordpiece tokenization。把每个token输入到wordpiece tokenizer, 然后把每个token的first sub-token作为input，输入到softmax里，其中h_n 是单词x_n对应的的first sub-token的hidden state.

![image](https://user-images.githubusercontent.com/10768193/56558733-9e064e00-65da-11e9-8e87-77f57373e820.png)

联合训练的目标函数是：

![image](https://user-images.githubusercontent.com/10768193/56558754-ab233d00-65da-11e9-8512-31c19bbef81e.png)

3.3 Conditional Random Field

另外，针对SF，有报告称使用CRF有提升效果，所以这篇文章也对这方面进行了调查。

**模型图：**

![image](https://user-images.githubusercontent.com/10768193/56558214-eb81bb80-65d8-11e9-9a66-b4bc9975d0d9.png)

[CLS]对应的标签是分类的标签。

**结果**：

ATIS dataset：training, development and test sets contain 4,478, 500 and 893 utterances. 120 slot labels and 21 intent types for the training set.

Snips: The training, development and test sets contain 13,084, 700 and 700 utterances, respectively. There are 72 slot labels and 7 intent types for the training set.

使用了，BERT-base，fine-tune，长度最长是50（这也太短了吧），batch size是128， adam， lr是5e-5. dropout 是0.1. The maximum number of epochs is selected from [1, 5, 10, 20, 30, 40].

等等，这篇文章里没有说是否使用了LML模型（mask）

![image](https://user-images.githubusercontent.com/10768193/56559118-9d21ec00-65db-11e9-9d29-be4395e9ba85.png)

![image](https://user-images.githubusercontent.com/10768193/56559209-e83bff00-65db-11e9-80b1-d3bcda9c86d1.png)

**接下来要看的论文**：
